File: lib/db/src/schema/organisations.ts
import { pgTable, serial, varchar, text, timestamp, boolean, integer, numeric, bigint } from "drizzle-orm/pg-core";

export const organisationsTable = pgTable("organisations", {
  id: serial("id").primaryKey(),
  name: varchar("name", { length: 200 }).notNull(),
  slug: varchar("slug", { length: 100 }).notNull().unique(),
  email: varchar("email", { length: 255 }),
  phone: varchar("phone", { length: 30 }),
  address: text("address"),
  logo: text("logo"),
  maxUsers: integer("max_users").notNull().default(5),
  actif: boolean("actif").notNull().default(true),
  bankName: varchar("bank_name", { length: 200 }),
  bankIban: varchar("bank_iban", { length: 50 }),
  bankBic: varchar("bank_bic", { length: 20 }),
  siret: varchar("siret", { length: 20 }),
  tvaNumber: varchar("tva_number", { length: 30 }),
  legalForm: varchar("legal_form", { length: 100 }),
  capital: varchar("capital", { length: 50 }),
  invoiceFooter: text("invoice_footer"),
  autoInvoiceEnabled: boolean("auto_invoice_enabled").notNull().default(true),
  autoEmailInvoice: boolean("auto_email_invoice").notNull().default(true),
  weeklySecurityEmail: boolean("weekly_security_email").notNull().default(false),
  lastSecurityDigestAt: timestamp("last_security_digest_at", { withTimezone: true }),
  proactiveEngineEnabled: boolean("proactive_engine_enabled").notNull().default(true),
  // Capture automatique des dépenses : tout justificatif entrant (upload ou
  // pièce jointe e-mail) reconnu comme facture/note de frais est analysé par
  // Document IA puis poussé dans la file d'inspection des dépenses. Voir
  // services/expense-capture.ts.
  expenseAutoCaptureEnabled: boolean("expense_auto_capture_enabled").notNull().default(true),
  // Réglages par org du moteur proactif (réglables depuis l'UI). Défauts =
  // valeurs historiques codées en dur (8 h / 21 j). Voir proactive-engine.ts.
  messageSlaHours: integer("message_sla_hours").notNull().default(8),
  quietCustomerAfterDays: integer("quiet_customer_after_days").notNull().default(21),
  // Relances de paiement automatiques (services/invoice-reminder-cron.ts).
  // Ces e-mails partent vers les CLIENTS de l'organisation: elle doit pouvoir
  // les couper entierement (avant: aucun moyen de le faire, le cron passait
  // sur toutes les organisations sans condition).
  autoRemindersEnabled: boolean("auto_reminders_enabled").notNull().default(true),
  // Quand true (defaut), les relances de paiement et les factures mensuelles
  // generees automatiquement passent par la file d'approbation au lieu d'etre
  // envoyees/finalisees directement. Ce sont des actions visibles par le
  // client final et difficiles a rattraper: un humain doit les voir d'abord.
  billingRequiresApproval: boolean("billing_requires_approval").notNull().default(true),
  agentAutoRunEnabled: boolean("agent_auto_run_enabled").notNull().default(false),
  agentAutoRunLastRunAt: timestamp("agent_auto_run_last_run_at", { withTimezone: true }),
  // Horaires d'ouverture utilises par le service de disponibilites (creneaux
  // de rendez-vous). Jours = numeros ISO (1=lundi .. 7=dimanche) separes par
  // des virgules. Heures = "HH:MM" 24h. Fuseau IANA. Duree par defaut d'un RDV.
  workingDays: varchar("working_days", { length: 20 }).notNull().default("1,2,3,4,5"),
  workingHoursStart: varchar("working_hours_start", { length: 5 }).notNull().default("09:00"),
  workingHoursEnd: varchar("working_hours_end", { length: 5 }).notNull().default("18:00"),
  appointmentTimezone: varchar("appointment_timezone", { length: 60 }).notNull().default("Europe/Paris"),
  appointmentDurationMinutes: integer("appointment_duration_minutes").notNull().default(30),
  aiLearningLastRunAt: timestamp("ai_learning_last_run_at", { withTimezone: true }),
  reusedScanCount: integer("reused_scan_count").notNull().default(0),
  reusedScanSavedMs: bigint("reused_scan_saved_ms", { mode: "number" }).notNull().default(0),
  aiQuotaCostUsd: numeric("ai_quota_cost_usd", { precision: 10, scale: 2 }),
  aiQuotaCalls: integer("ai_quota_calls"),
  aiAgentName: varchar("ai_agent_name", { length: 100 }),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow().$onUpdate(() => new Date()),
});

export type Organisation = typeof organisationsTable.$inferSelect;
export type InsertOrganisation = typeof organisationsTable.$inferInsert;

File: apps/api-py/src/agent_bureau_api/db/models/organisations.py
"""1:1 port of lib/db/src/schema/organisations.ts. The tenant root — not
itself tenant-scoped (it IS the tenant)."""
from __future__ import annotations

from datetime import datetime

from sqlalchemy import Boolean, Integer, Numeric, String, Text, BigInteger
from sqlalchemy.orm import Mapped, mapped_column

from ..base import Base
from ..mixins import TimestampMixin


class Organisation(Base, TimestampMixin):
    __tablename__ = "organisations"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(100), nullable=False, unique=True)
    email: Mapped[str | None] = mapped_column(String(255))
    phone: Mapped[str | None] = mapped_column(String(30))
    address: Mapped[str | None] = mapped_column(Text)
    logo: Mapped[str | None] = mapped_column(Text)
    max_users: Mapped[int] = mapped_column(Integer, nullable=False, default=5, server_default="5")
    actif: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")

    bank_name: Mapped[str | None] = mapped_column(String(200))
    bank_iban: Mapped[str | None] = mapped_column(String(50))
    bank_bic: Mapped[str | None] = mapped_column(String(20))
    siret: Mapped[str | None] = mapped_column(String(20))
    tva_number: Mapped[str | None] = mapped_column(String(30))
    legal_form: Mapped[str | None] = mapped_column(String(100))
    capital: Mapped[str | None] = mapped_column(String(50))
    invoice_footer: Mapped[str | None] = mapped_column(Text)

    auto_invoice_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    auto_email_invoice: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    weekly_security_email: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")
    last_security_digest_at: Mapped[datetime | None]
    proactive_engine_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    expense_auto_capture_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")

    message_sla_hours: Mapped[int] = mapped_column(Integer, nullable=False, default=8, server_default="8")
    quiet_customer_after_days: Mapped[int] = mapped_column(Integer, nullable=False, default=21, server_default="21")
    agent_auto_run_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")
    agent_auto_run_last_run_at: Mapped[datetime | None]

    working_days: Mapped[str] = mapped_column(String(20), nullable=False, default="1,2,3,4,5", server_default="1,2,3,4,5")
    working_hours_start: Mapped[str] = mapped_column(String(5), nullable=False, default="09:00", server_default="09:00")
    working_hours_end: Mapped[str] = mapped_column(String(5), nullable=False, default="18:00", server_default="18:00")
    appointment_timezone: Mapped[str] = mapped_column(String(60), nullable=False, default="Europe/Paris", server_default="Europe/Paris")
    appointment_duration_minutes: Mapped[int] = mapped_column(Integer, nullable=False, default=30, server_default="30")

    ai_learning_last_run_at: Mapped[datetime | None]
    reused_scan_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    reused_scan_saved_ms: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0, server_default="0")
    ai_quota_cost_usd: Mapped[float | None] = mapped_column(Numeric(10, 2))
    ai_quota_calls: Mapped[int | None] = mapped_column(Integer)
    ai_agent_name: Mapped[str | None] = mapped_column(String(100))
